tfjs-backend-webgpu/src/kernels/reduce_webgpu.ts

```diff
@@ -0,0 +1,129 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * =============================================================================
+ */
+
+import {backend_util, util} from '@tensorflow/tfjs-core';
+import {ReduceInfo} from '../../../tfjs-core/src/ops/reduce_util';
+import {getCoordsDataType} from '../shader_preprocessor';
+import {computeDispatch} from '../webgpu_util';
+import {WebGPUProgram} from './webgpu_program';
+
+export class ReduceProgram implements WebGPUProgram {
+  outputShape: number[];
+  userCode: string;
+  dispatchLayout: {x: number[], y: number[]};
+  dispatch: [number, number, number];
+  workGroupSize: [number, number, number];
+  variableNames = ['x'];
+
+  constructor(reduceInfo: ReduceInfo, reduceType: 'max'|'min'|'sum') {
+    const inputShape = [reduceInfo.batchSize, reduceInfo.inSize];
+    const [outputShape, reduceShape] =
+        backend_util.computeOutAndReduceShapes(inputShape, [1]);
+    this.outputShape = outputShape.length === 0 ? [1] : outputShape;
+    const reduceSize = util.sizeFromShape(reduceShape);
+
+    const op = reduceType === 'min' ? '<' : '>';
+    const reductionFactor = 2;
+    const xMaxThreads = 1024;
+    const xThreads =
+        Math.min(Math.ceil(reduceSize / reductionFactor), xMaxThreads);
+
+    this.workGroupSize = [xThreads, 1, 1];
+    this.dispatchLayout = {x: [], y: this.outputShape.map((d, i) => i)};
+    this.dispatch = computeDispatch(this.dispatchLayout, this.outputShape);
+    const reduceInSharedMemory = xThreads > 1;
+
+    const minmaxOp = `
+          if (candidate ${op} bestValue && !isnan(candidate)) {
+            bestValue = candidate;
+          }
+      `;
+    const sumOp = `
+            bestValue += candidate;
+      `;
+    const sharedMemorySnippet = `
+        shared float xBestValues[WorkGroupSize];
+      `;
+    const sharedMemoryReduceSnippet = `
+      xBestValues[gl_LocalInvocationID.x] = bestValue;
+      ${reduceType === 'sum' ? `bestValue=0;` : ``}
+      int currentSize = WorkGroupSize;
+      while (currentSize > 1) {
+        barrier();
+        for (int w = 0; w < ${reductionFactor}; ++w) {
+          int i = int(gl_LocalInvocationID.x) * ${reductionFactor} + w;
+          if (i < currentSize) {
+            float candidate = xBestValues[i];
+            ${(reduceType === 'min' || reduceType === 'max') ? minmaxOp : sumOp}
+          }
+        }
+        xBestValues[gl_LocalInvocationID.x] = bestValue;
+        currentSize = DIV_CEIL(currentSize, ${reductionFactor});
+        ${reduceType === 'sum' ? `if(currentSize > 1) bestValue=0;` : ``}
+      }
+      if (gl_LocalInvocationID.x == 0) {
+        setOutput(flatOutputIndex, bestValue);
+      }
+    `;
+
+    const outputCoordsType = getCoordsDataType(this.outputShape.length);
+    const indexOutputCoords = (outputCoords: string, index: string) => {
```

Comment:
i think this and `indexInputShape` are simple enough that they should be inlined

something like `this.outputShape.length === 1 ? 'outputCoords' : 'outputCoords[0]`

to further simplify you could define a const `outputRank`